tutorials/backend/hasura-v3-ts-connector/tutorial-site/content/predicates.md

```diff
@@ -0,0 +1,352 @@
+            "name": "artist_id",
+            "path": []
+        "operator"; {
+            "type": "equal"
+        ｝，
+        "value"; {
+            "type": "scalar",
+            "value": 5  
+        }
+    }
+}
+// ...
+```
+
+This predicate expression has type `binary_comparison_operator`, which means it is a predicate which compares a column
+to a value using an operator - in this case, the equality operator - this predicate asserts that the `artist_id` column
+equals the literal value `5`.
+
+## Expression Types
+
+In the SDK, these predicate expressions are given the TypeScript type `Expression`, and we can see that there are
+several different types of expression.
+
+```typescript
+export type Expression = {
+    expressions: Expression[];
+    type: "and";
+} | {
+    expressions: Expression[];
+    type: "or";
+} | {
+    expression: Expression;
+    type: "not";
+} | {
+    column: ComparisonTarget;
+    operator: UnaryComparisonOperator;
+    type: "unary_comparison_operator";
+} | {
+    column: ComparisonTarget;
+    operator: BinaryComparisonOperator;
+    type: "binary_comparison_operator";
+    value: ComparisonValue;
+} | {
+    column: ComparisonTarget;
+    operator: BinaryArrayComparisonOperator;
+    type: "binary_array_comparison_operator";
+    values: ComparisonValue[];
+} | {
+    in_collection: ExistsInCollection;
+    type: "exists";
+    where: Expression;
+};
+```
+
+There are logical expressions like `and`, `or`, and `not`, which combine other simpler expressions.
+
+There are unary and binary comparison operator expressions.
+
+And there are `exists` expressions, which are expressed using a subquery against another collection.
+
+For now, we'll concentrate on logical expressions and comparison operator expressions.
+
+We're going to build up the `WHERE` clause recursively, starting with the simplest expressions at the leaves of the
+predicate expression tree, and working upwards. As we go, we will need to keep track of any query parameters that we
+also need to pass to SQLite, so let's make a place to store those.
+
+In the `fetch_rows` function, let's add a new variable to store our parameters:
+
+```typescript
+const parameters: any[] = [];
+```
+
+As we encounter literal values, we'll append them as new parameters to this list.
+
+We also need to add our parameters to the database fetch function call, and to our logging function:
+
+```typescript
+console.log(JSON.stringify({ sql, parameters }));
+
+return state.db.all(sql, ...parameters);
+```
+
+Let's delegate to a helper function in order to build our `WHERE` clause. Let's call our function `visit_expression`,
+because we're using the visitor design pattern.
+
+```typescript
+const where_clause = request.query.where == null ? "" : `WHERE ${visit_expression(parameters, request.query.where)}`;
+```
+
+We'll handle each different type of expression by pattern matching on the `type` field of the current expression.
+
+```typescript
```

Comment:
As we're adding these snippets without context, a solution to help a user understand where they should add these might be to link to their lines on the completed `main` branch [like this](https://github.com/hasura/ndc-typescript-learn-course/blob/main/src/index.ts#L218).

Reply:
I've solved this in content